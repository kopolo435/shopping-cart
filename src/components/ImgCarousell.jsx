import React from "react";
import PropTypes from "prop-types";
import Button from "./Button";

function ImgCarousell({ imgList }) {
  const [currentImgIndex, setCurrentImgIndex] = React.useState(0);

  function moveToRight() {
    if (currentImgIndex === imgList.length - 1) {
      setCurrentImgIndex(0);
    } else {
      setCurrentImgIndex((currentIndex) => currentIndex + 1);
    }
  }

  function moveToLeft() {
    if (currentImgIndex === 0) {
      setCurrentImgIndex(imgList.length - 1);
    } else {
      setCurrentImgIndex((currentIndex) => currentIndex - 1);
    }
  }

  return (
    <div className="imgCarousell">
      <div className="carousellTop">
        <Button
          type="button"
          onClick={moveToLeft}
          className="carousellBtn"
          label="show previous image"
        >
          left
        </Button>
        <div className="carousellContent">
          <div className="imgContainer">
            {imgList.map((imgData, index) => {
              if (index === currentImgIndex) {
                return (
                  <img
                    src={imgData.src}
                    key={index}
                    alt={imgData.alt}
                    data-testid="currentImg"
                  />
                );
              }

              return <img key={index} src={imgData.src} alt={imgData.alt} />;
            })}
          </div>
          <h3 data-testid="currentImgTitle">
            {imgList[currentImgIndex].title}
          </h3>
        </div>
        <Button
          type="button"
          onClick={moveToRight}
          className="carousellBtn"
          label="show next image"
        >
          left
        </Button>
      </div>
      <div className="imgControl">
        {imgList.map((data, index) => {
          if (currentImgIndex === index) {
            return (
              <Button
                type="button"
                content=""
                className="chooseImg"
                onClick={() => setCurrentImgIndex(index)}
                label={`show image ${index}`}
                key={index}
              />
            );
          }
          return (
            <Button
              type="button"
              className="chooseImg"
              onClick={() => setCurrentImgIndex(index)}
              label={`show image ${index}`}
              key={index}
            />
          );
        })}
      </div>
    </div>
  );
}

ImgCarousell.propTypes = {
  imgList: PropTypes.array.isRequired,
};

export default ImgCarousell;
