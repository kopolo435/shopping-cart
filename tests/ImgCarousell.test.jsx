import { render, screen } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import React from "react";
import userEvent from "@testing-library/user-event";
import ImgCarousell from "../src/components/ImgCarousell";

it("Carousell renders right", () => {
  const imgData = [
    { title: "img1", src: "url/sitio1", alt: "Imagen 1" },
    { title: "img2", src: "url/sitio2", alt: "Imagen 2" },
    { title: "img3", src: "url/sitio3", alt: "Imagen 3" },
  ];

  const { container } = render(<ImgCarousell imgList={imgData} />);
  expect(container).toMatchSnapshot();
});

it("Shows correct img", () => {
  const imgData = [
    { title: "img1", src: "./url/sitio1", alt: "Imagen 1" },
    { title: "img2", src: "./url/sitio2", alt: "Imagen 2" },
    { title: "img3", src: "./url/sitio3", alt: "Imagen 3" },
  ];
  render(<ImgCarousell imgList={imgData} />);

  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio1"
  );
  expect(screen.getByTestId("currentImgTitle").textContent).toBe("img1");
});

it("Move to the right button shows correct img", async () => {
  const imgData = [
    { title: "img1", src: "http://localhost:3000/url/sitio1", alt: "Imagen 1" },
    { title: "img2", src: "http://localhost:3000/url/sitio2", alt: "Imagen 2" },
    { title: "img3", src: "http://localhost:3000/url/sitio3", alt: "Imagen 3" },
  ];
  const user = userEvent.setup();
  render(<ImgCarousell imgList={imgData} />);

  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio1"
  );

  const button = screen.getByRole("button", { name: "show next image" });
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio2"
  );
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio3"
  );
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio1"
  );
});

it("Move to the left button shows correct img", async () => {
  const imgData = [
    { title: "img1", src: "http://localhost:3000/url/sitio1", alt: "Imagen 1" },
    { title: "img2", src: "http://localhost:3000/url/sitio2", alt: "Imagen 2" },
    { title: "img3", src: "http://localhost:3000/url/sitio3", alt: "Imagen 3" },
  ];
  const user = userEvent.setup();
  render(<ImgCarousell imgList={imgData} />);

  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio1"
  );

  const button = screen.getByRole("button", { name: "show previous image" });
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio3"
  );
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio2"
  );
  await user.click(button);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio1"
  );
});

it("Choose img button show correct img", async () => {
  const imgData = [
    { title: "img1", src: "./url/sitio1", alt: "Imagen 1" },
    { title: "img2", src: "./url/sitio2", alt: "Imagen 2" },
    { title: "img3", src: "./url/sitio3", alt: "Imagen 3" },
  ];
  const user = userEvent.setup();
  render(<ImgCarousell imgList={imgData} />);

  const showImgButtons = screen.getAllByRole("button", {
    name: /.+[\d]+/im,
  });

  await user.click(showImgButtons[2]);
  expect(screen.getByTestId("currentImg").src).toBe(
    "http://localhost:3000/url/sitio3"
  );
  expect(screen.getByTestId("currentImgTitle").textContent).toBe("img3");
});
