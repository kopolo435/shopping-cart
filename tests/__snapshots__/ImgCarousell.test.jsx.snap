// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Carousell renders right 1`] = `
<div>
  <div
    class="imgCarousell"
  >
    <div
      class="carousellTop"
    >
      <button
        aria-label="show previous image"
        class="carousellBtn"
        type="button"
      >
        left
      </button>
      <div
        class="carousellContent"
      >
        <div
          class="imgContainer"
        >
          <img
            alt="Imagen 1"
            data-testid="currentImg"
            src="url/sitio1"
          />
          <img
            alt="Imagen 2"
            src="url/sitio2"
          />
          <img
            alt="Imagen 3"
            src="url/sitio3"
          />
        </div>
        <h3
          data-testid="currentImgTitle"
        >
          img1
        </h3>
      </div>
      <button
        aria-label="show next image"
        class="carousellBtn"
        type="button"
      >
        left
      </button>
    </div>
    <div
      class="imgControl"
    >
      <button
        aria-label="show image 0"
        class="chooseImg"
        type="button"
      />
      <button
        aria-label="show image 1"
        class="chooseImg"
        type="button"
      />
      <button
        aria-label="show image 2"
        class="chooseImg"
        type="button"
      />
    </div>
  </div>
</div>
`;
